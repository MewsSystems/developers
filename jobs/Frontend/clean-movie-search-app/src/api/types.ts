export interface Movie {
  id: number;
  title: string;
  overview: string;
  poster_path: string | null;
  release_date: string;
  vote_average: number;
}

export interface MovieSearchResponse {
  page: number;
  results: Movie[];
  total_pages: number;
  total_results: number;
}
