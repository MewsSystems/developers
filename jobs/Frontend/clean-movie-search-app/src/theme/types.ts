export interface Theme {
  name: 'light' | 'dark';
  colors: {
    primary: string;
    secondary: string;
    background: string;
    surface: string;
    text: {
      primary: string;
      secondary: string;
    };
    button: {
      background: string;
      hover: string;
      text: string;
      inverseText: string;
    };
    card: {
      background: string;
      shadow: string;
      hoverShadow: string;
    };
    rating: {
      background: string;
      text: string;
      highRating: {
        border: string;
        text: string;
      };
    };
    input: {
      background: string;
      border: string;
      focusBorder: string;
    };
  };
  spacing: {
    xs: string;
    sm: string;
    md: string;
    lg: string;
    xl: string;
  };
  borderRadius: {
    sm: string;
    md: string;
    lg: string;
  };
}
