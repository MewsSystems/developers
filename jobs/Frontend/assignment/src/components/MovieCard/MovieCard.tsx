import styled from "styled-components";
import { Chip, LinkButton, Rating, Typography } from "..";
import fallbackImg from "@/assets/mocks/fallback.jpg";
import { Movie } from "tmdb-ts";
import { MEDIA_300_BASE_URL } from "@/tmdbClient";

export interface MovieCardProps {
  movie: Movie;
  genres?: string[];
}

const CardWrapper = styled.div`
  display: flex;
  flex-direction: column;

  max-height: 540px;
  min-width: 260px;
  width: 100%;

  border-radius: 16px;
  overflow: hidden;
  border: 1px solid ${({ theme }) => theme.colors.outline.variant};
`;

const CardImage = styled.img`
  height: 310px;
  object-fit: cover;
`;

const CardContent = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  gap: 16px;

  flex: 1;
  max-height: 100%;
  overflow: hidden;

  padding: 16px;
`;

const TextWrapper = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  gap: 8px;

  flex: 1;
  max-height: 75%;
  overflow: hidden;

  height: 100%;
`;

const DescriptionWrapper = styled.div`
  display: -webkit-box;
  -webkit-line-clamp: 3;
  -webkit-box-orient: vertical;
  overflow: hidden;
  flex: 1;
  max-height: 35%;
  overflow: hidden;
`;

const ChipsWrapper = styled.div`
  display: flex;
  flex-wrap: wrap;
  gap: 8px;
`;

const ActionsWrapper = styled.div`
  display: flex;
  gap: 8px;
  justify-content: space-between;
  align-items: center;
`;

export function MovieCard({ movie, genres }: MovieCardProps) {
  const {
    id,
    title,
    overview,
    release_date: releaseDate,
    vote_average,
    poster_path: posterPath,
  } = movie;

  return (
    <CardWrapper>
      <CardImage
        src={posterPath ? MEDIA_300_BASE_URL + posterPath : fallbackImg}
        alt="Movie poster"
      />
      <CardContent>
        <TextWrapper>
          <Typography variant="titleSmall">{title}</Typography>
          <Typography variant="bodySmall" color="secondary">
            Release date: {releaseDate || "Unknown"}
          </Typography>
          <DescriptionWrapper>
            <Typography variant="bodySmall">{overview}</Typography>
          </DescriptionWrapper>
          <ChipsWrapper>
            {genres?.map((genre, index) => <Chip key={genre + index} label={genre} />)}
          </ChipsWrapper>
        </TextWrapper>
        <ActionsWrapper>
          <Rating value={vote_average / 2} />
          <LinkButton href={`/movie/${id}`}>See more</LinkButton>
        </ActionsWrapper>
      </CardContent>
    </CardWrapper>
  );
}
