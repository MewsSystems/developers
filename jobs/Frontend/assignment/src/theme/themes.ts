import { DefaultTheme } from "styled-components";
import { baseTheme, colorPalette } from "./base-theme";

export const materialLightTheme: DefaultTheme = {
  ...baseTheme,
  colors: {
    ...baseTheme.colors,
    primary: {
      main: colorPalette.primary[40],
      on: colorPalette.primary[100],
      container: colorPalette.primary[90],
      onContainer: colorPalette.primary[10],
      fixed: colorPalette.primary[90],
      onFixed: colorPalette.primary[10],
      surface: colorPalette.primary[100],
      onSurface: colorPalette.primary[100],
    },
    secondary: {
      main: colorPalette.secondary[40],
      on: colorPalette.secondary[100],
      container: colorPalette.secondary[90],
      onContainer: colorPalette.secondary[10],
      fixed: colorPalette.secondary[90],
      onFixed: colorPalette.secondary[10],
      surface: colorPalette.secondary[100],
      onSurface: colorPalette.secondary[100],
    },
    tertiary: {
      main: colorPalette.tertiary[40],
      on: colorPalette.tertiary[100],
      container: colorPalette.tertiary[90],
      onContainer: colorPalette.tertiary[10],
      fixed: colorPalette.tertiary[90],
      onFixed: colorPalette.tertiary[10],
      surface: colorPalette.tertiary[100],
      onSurface: colorPalette.tertiary[100],
    },
    error: {
      main: colorPalette.error[40],
      on: colorPalette.error[100],
      container: colorPalette.error[90],
      onContainer: colorPalette.error[10],
    },
  },
};

export const materialDarkTheme: DefaultTheme = {
  ...baseTheme,
  colors: {
    ...baseTheme.colors,
    primary: {
      main: colorPalette.primary[80],
      on: colorPalette.primary[20],
      container: colorPalette.primary[30],
      onContainer: colorPalette.primary[90],
      fixed: colorPalette.primary[90],
      onFixed: colorPalette.primary[10],
      surface: colorPalette.primary[100],
      onSurface: colorPalette.primary[100],
    },
    secondary: {
      main: colorPalette.secondary[80],
      on: colorPalette.secondary[20],
      container: colorPalette.secondary[30],
      onContainer: colorPalette.secondary[90],
      fixed: colorPalette.secondary[90],
      onFixed: colorPalette.secondary[10],
      surface: colorPalette.secondary[100],
      onSurface: colorPalette.secondary[100],
    },
    tertiary: {
      main: colorPalette.tertiary[80],
      on: colorPalette.tertiary[20],
      container: colorPalette.tertiary[30],
      onContainer: colorPalette.tertiary[90],
      fixed: colorPalette.tertiary[90],
      onFixed: colorPalette.tertiary[10],
      surface: colorPalette.tertiary[100],
      onSurface: colorPalette.tertiary[100],
    },
    error: {
      main: colorPalette.error[80],
      on: colorPalette.error[20],
      container: colorPalette.error[30],
      onContainer: colorPalette.error[90],
    },
  },
};
