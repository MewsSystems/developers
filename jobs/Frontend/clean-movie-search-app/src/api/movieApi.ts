import { Movie, MovieSearchResponse } from './types';

const API_KEY = '03b8572954325680265531140190fd2a';
const BASE_URL = 'https://api.themoviedb.org/3';

export const searchMovies = async (
  query: string,
  page: number = 1
): Promise<MovieSearchResponse> => {
  // Introduce a delay of 0.5 second (500 milliseconds) - Kayak style
  await new Promise((resolve) => setTimeout(resolve, 500));

  const response = await fetch(
    `${BASE_URL}/search/movie?api_key=${API_KEY}&query=${encodeURIComponent(query)}&page=${page}`
  );
  if (!response.ok) {
    throw new Error('Failed to fetch movies');
  }
  return response.json();
};

export const getMovieDetails = async (movieId: number): Promise<Movie> => {
  const response = await fetch(
    `${BASE_URL}/movie/${movieId}?api_key=${API_KEY}`
  );
  if (!response.ok) {
    throw new Error('Failed to fetch movie details');
  }
  return response.json();
};
