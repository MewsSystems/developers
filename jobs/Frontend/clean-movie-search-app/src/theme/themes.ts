import { Theme } from './types';


export const darkTheme: Theme = {
    name: 'dark',
    colors: {
      primary: '#4db6ac',
      secondary: '#80cbc4',
      background: '#121212',
      surface: '#1e1e1e',
      text: {
        primary: '#ffffff',
        secondary: '#b3b3b3',
      },
      button: {
        background: '#4db6ac',
        hover: '#80cbc4',
        text: '#000000',
        inverseText: '#ffffff',
      },
      card: {
        background: '#1e1e1e',
        shadow: '0 2px 8px rgba(0, 0, 0, 0.2)',
        hoverShadow: '0 4px 16px rgba(0, 0, 0, 0.4)',
      },
      rating: {
        background: 'rgba(0, 0, 0, 0.75)',
        text: '#ffffff',
        highRating: {
          border: '#ffd700',
          text: '#ffd700',
        },
      },
      input: {
        background: '#2d2d2d',
        border: '#404040',
        focusBorder: '#4db6ac',
      },
    },
    spacing: {
      xs: '4px',
      sm: '8px',
      md: '16px',
      lg: '24px',
      xl: '32px',
    },
    borderRadius: {
      sm: '4px',
      md: '8px',
      lg: '16px',
    },
  };
  
  export const lightTheme: Theme = {
    name: 'light',
    colors: {
      primary: '#26a69a',
      secondary: '#4db6ac',
      background: '#f5f5f5',
      surface: '#ffffff',
      text: {
        primary: '#333333',
        secondary: '#666666',
      },
      button: {
        background: '#26a69a',
        hover: '#4db6ac',
        text: '#ffffff',
        inverseText: '#000000',
      },
      card: {
        background: '#ffffff',
        shadow: '0 2px 8px rgba(0, 0, 0, 0.1)',
        hoverShadow: '0 4px 16px rgba(0, 0, 0, 0.2)',
      },
      rating: {
        background: 'rgba(0, 0, 0, 0.75)',
        text: '#ffffff',
        highRating: {
          border: '#ffd700',
          text: '#ffd700',
        },
      },
      input: {
        background: '#ffffff',
        border: '#e0e0e0',
        focusBorder: '#26a69a',
      },
    },
    spacing: {
      xs: '4px',
      sm: '8px',
      md: '16px',
      lg: '24px',
      xl: '32px',
    },
    borderRadius: {
      sm: '4px',
      md: '8px',
      lg: '16px',
    },
  };
