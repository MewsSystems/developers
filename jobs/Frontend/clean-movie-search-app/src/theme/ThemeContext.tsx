import React, { createContext, useContext, useState, useCallback } from 'react';
import { ThemeProvider as StyledThemeProvider } from 'styled-components';
import { Theme } from './types';
import { lightTheme, darkTheme } from './themes';

interface ThemeContextType {
  theme: Theme;
  toggleTheme: () => void;
}

const ThemeContext = createContext<ThemeContextType | undefined>(undefined);

export const ThemeProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  const [theme, setTheme] = useState<Theme>(darkTheme);

  const toggleTheme = useCallback(() => {
    setTheme((current) => (current.name === 'light' ? darkTheme : lightTheme));
  }, []);

  return (
    <ThemeContext.Provider value={{ theme, toggleTheme }}>
      <StyledThemeProvider theme={theme}>{children}</StyledThemeProvider>
    </ThemeContext.Provider>
  );
};

export const useTheme = () => {
  const context = useContext(ThemeContext);
  if (context === undefined) {
    throw new Error('useTheme must be used within a ThemeProvider');
  }
  return context;
};
