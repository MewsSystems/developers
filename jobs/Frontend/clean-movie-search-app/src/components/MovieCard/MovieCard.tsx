import React from 'react';
import styled, { css } from 'styled-components';
import { Movie } from '../../api';

const Card = styled.article`
  border-radius: ${({ theme }) => theme.borderRadius.md};
  overflow: hidden;
  background: ${({ theme }) => theme.colors.card.background};
  box-shadow: ${({ theme }) => theme.colors.card.shadow};
  transition:
    transform 0.2s,
    box-shadow 0.2s,
    background-color 0.3s;
  cursor: pointer;

  &:hover {
    transform: translateY(-2px);
    box-shadow: ${({ theme }) => theme.colors.card.hoverShadow};
  }
`;

const PosterContainer = styled.div`
  position: relative;
  padding-top: 150%;
`;

const Poster = styled.img`
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  object-fit: cover;
`;

const Content = styled.div`
  padding: ${({ theme }) => theme.spacing.md};
`;

const Title = styled.h3`
  margin: 0 0 ${({ theme }) => theme.spacing.sm} 0;
  font-size: 1rem;
  font-weight: 600;
  color: ${({ theme }) => theme.colors.text.primary};
`;

const ReleaseDate = styled.p`
  margin: 0;
  color: ${({ theme }) => theme.colors.text.secondary};
  font-size: 0.875rem;
`;

const Rating = styled.div<{ rating: number }>`
  position: absolute;
  top: ${({ theme }) => theme.spacing.sm};
  left: ${({ theme }) => theme.spacing.sm};
  background: ${({ theme }) => theme.colors.rating.background};
  color: ${({ theme }) => theme.colors.rating.text};
  padding: ${({ theme }) => `${theme.spacing.xs} ${theme.spacing.sm}`};
  border-radius: ${({ theme }) => theme.borderRadius.sm};
  font-size: 0.875rem;

  ${({ rating, theme }) =>
    rating >= 7.5 &&
    css`
      color: ${theme.colors.rating.highRating.text};
      border: 2px solid ${theme.colors.rating.highRating.border};
    `}
`;

interface MovieCardProps {
  movie: Movie;
  onClick: (id: number) => void;
}

export const MovieCard: React.FC<MovieCardProps> = ({ movie, onClick }) => {
  const releaseDate = new Date(movie.release_date);
  const isValidDate = !isNaN(releaseDate.getDate());

  return (
    <Card onClick={() => onClick(movie.id)} aria-label={`Movie card for ${movie.title}`}> 
      <PosterContainer>
        {movie.poster_path ? (
          <Poster
            src={`https://image.tmdb.org/t/p/w500${movie.poster_path}`}
            alt={movie.title}
            loading="lazy"
          />
        ) : (
          <Poster
            src={`https://via.placeholder.com/500x750?text=${encodeURIComponent(
              movie.title
            )}`}
            alt={movie.title}
          />
        )}
        {typeof movie.vote_average === 'number' && ( 
          <Rating rating={movie.vote_average} aria-label={`Rating: ${movie.vote_average.toFixed(1)}`}> 
            {movie.vote_average.toFixed(1)}
          </Rating>
        )}
      </PosterContainer>
      <Content>
        <Title>{movie.title}</Title>
        {isValidDate && (
          <ReleaseDate>{releaseDate.toLocaleDateString()}</ReleaseDate>
        )}
      </Content>
    </Card>
  );
};