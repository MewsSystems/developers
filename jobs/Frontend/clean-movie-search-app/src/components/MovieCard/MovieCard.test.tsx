import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { lightTheme } from '../../theme/themes';
import { MovieCard } from './MovieCard';
import { Movie } from '../../api';

describe('MovieCard', () => {
  const mockMovie: Movie = {
    id: 1,
    title: 'Movie Title',
    overview: 'Movie overview',
    poster_path: '/poster.jpg',
    release_date: '2024-11-17',
    vote_average: 7.8,
  };

  it('renders the movie card correctly', () => {
    render(
      <ThemeProvider theme={lightTheme}>
        <MovieCard movie={mockMovie} onClick={() => {}} />
      </ThemeProvider>
    );

    expect(screen.getByRole('img', { name: 'Movie Title' })).toHaveAttribute(
      'src',
      'https://image.tmdb.org/t/p/w500/poster.jpg'
    );
    expect(screen.getByText('Movie Title')).toBeInTheDocument();
    expect(screen.getByText('11/17/2024')).toBeInTheDocument();
    expect(screen.getByText('7.8')).toBeInTheDocument();
  });

  it('renders a placeholder image if poster_path is missing', () => {
    render(
      <ThemeProvider theme={lightTheme}>
        <MovieCard
          movie={{ ...mockMovie, poster_path: null }}
          onClick={() => {}}
        />
      </ThemeProvider>
    );

    expect(screen.getByRole('img', { name: 'Movie Title' })).toHaveAttribute(
      'src',
      'https://via.placeholder.com/500x750?text=Movie%20Title'
    );
  });

  it('renders the correct rating color based on vote_average', () => {
    render(
      <ThemeProvider theme={lightTheme}>
        <MovieCard movie={mockMovie} onClick={() => {}} />
      </ThemeProvider>
    );

    const ratingElement = screen.getByText('7.8');
    expect(ratingElement).toHaveStyle(
      `color: ${lightTheme.colors.rating.highRating.text}; border: 2px solid ${lightTheme.colors.rating.highRating.border}`
    );
  });

  it('matches the snapshot', () => {
    const { container } = render(
      <ThemeProvider theme={lightTheme}>
        <MovieCard movie={mockMovie} onClick={() => {}} />
      </ThemeProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it('calls onClick with the correct movie ID when clicked', () => {
    const mockOnClick = jest.fn();
    render(
      <ThemeProvider theme={lightTheme}>
        <MovieCard movie={mockMovie} onClick={mockOnClick} />
      </ThemeProvider>
    );

    const cardElement = screen.getByRole('article');
    fireEvent.click(cardElement);

    expect(mockOnClick).toHaveBeenCalledWith(mockMovie.id);
  });
});
