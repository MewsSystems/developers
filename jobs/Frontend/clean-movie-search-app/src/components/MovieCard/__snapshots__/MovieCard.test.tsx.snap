// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard matches the snapshot 1`] = `
<div>
  <article
    aria-label="Movie card for Movie Title"
    class="sc-blHHSb cFyGuL movie-card"
  >
    <div
      class="sc-gtLWhw ivqwil poster-container"
    >
      <img
        alt="Movie Title"
        class="sc-egkSDF dZwIUo poster"
        loading="lazy"
        src="https://image.tmdb.org/t/p/w500/poster.jpg"
      />
      <div
        aria-label="Rating: 7.8"
        class="sc-ghWlax jpWtuz rating"
      >
        7.8
      </div>
    </div>
    <div
      class="sc-fAUdSK cDEqUf content"
    >
      <h3
        class="sc-dntaoT fhBgiP title"
      >
        Movie Title
      </h3>
      <p
        class="sc-ivxoEo cffOek release-date"
      >
        11/17/2024
      </p>
    </div>
  </article>
</div>
`;
